### DISCUSSION TITLE
ZEPETOScript에서 namespace는 어떻게 사용하나요?

### DISCUSSION THREAD

**Discussion by mayko1234** [naverz/zepeto-studio-kor#289] [5. Q&A]
안녕하세요,
ZEPETOScript에서 namespace는 어떻게 사용하나요?
기존 타입스크립트에서 쓰던 방식으로 사용하려하니 컴파일 에러가 나더라구요.

**Comment by mayko1234** [comment]
네임스페이스를 넣고 default 키워드를 빼니까 컴파일 단계에서는 에러가 안나는데, 플레이 했을 시 첨부 스크린샷같은 에러가 보입니다. 클래스 선언시 export default 키워드가 필수인가요? 만약 필수라면 왜 그런지 아시는 분 있으면 리플 부탁드립니다~!
![image](https://user-images.githubusercontent.com/97988822/151460348-fcaa22f3-50cd-48c2-9076-f5a87f8bcb08.png)

**Reply by x4cf** [reply]
코드를 볼 수 있다면 같이 고민해보기 더 좋을것같아요 XD

**Reply by mayko1234** [reply]
안녕하세요~ 리플 감사합니다! 아 그렇네요 제가 왜 코드를 같이 안올렸을까요..
코드와 코드로 인해 발생한 에러 공유하겠습니다. 

![image](https://user-images.githubusercontent.com/97988822/151514598-17681975-250b-43db-8513-54f3d5abad5d.png)
![image](https://user-images.githubusercontent.com/97988822/151514683-b6cf1c40-05d6-4f6b-96a1-8f9192229824.png)

![image](https://user-images.githubusercontent.com/97988822/151514771-a22a4e5d-b58d-4b86-ae9d-dba23bc2dca9.png)
![image](https://user-images.githubusercontent.com/97988822/151514938-a5bce018-0af7-42e9-821f-297cebd203d2.png)

**Reply by x4cf** [reply]
와우... 처음보는 형태의 코드군요!
제페토 월드의 스크립트를 타입스크립트로 생성하고는 있지만, 제페토 월드의 스크립트 문법은 유니티를 결합하면서 기존의 타입스크립트 문법과는 조금 다른것으로 보여요.
아래와 같은 문법을 사용해보는건 어떨까요?
export default class "클래스 네임" extends ZepetoScriptBehaviour

**Reply by mayko1234** [reply]
넵 리플 감사합니다~ 좋은 하루 되세요~!

**Comment by runasy-koonta** [comment]
Zepeto Script를 GameObject에 포함시킬 때에는 default export된 class가 존재해야 합니다.
그런데 namespace된 class를 export default 할 수 없으므로 (export namespace가 가능하긴 하겠지만 마찬가지로 class를 export 하는 것은 아닙니다.)
아래와 같이 코드를 작성해야 할 것으로 보입니다.

```
namespace TestNamespace {
    export class TestClass extends ZepetoScriptBehaviour {
        ...codes
    }
}
export default TestNamespace.TestClass;
```

그런데 문제는 ZepetoScript compiler 문제인지 export default class [ClassName] 형태로 작성하지 않으면 (즉, class를 생성하면서 같이 export default 해주지 않으면) Inspector 창에서 public variable 에 접근할 수 없더라고요.

이 부분 감안하시고 사용하셔야 할 것 같습니다.

**Reply by runasy-koonta** [reply]
추가로, ZepetoScript가 쓰고 있는 것으로 추정되는 [puerts](https://github.com/Tencent/puerts)의 사용 코드들을 살펴보면, (https://github.com/zombieyang/puerts-wxengine-example/blob/master/Typescript/src/PhyWorld.ts)

대부분 위에 제가 적은 것과 같이 export default class 형태로 사용하고 있습니다.
아마 puerts의 한계가 아닌가 싶기도 합니다.

**Reply by mayko1234** [reply]
안녕하세요, 
말씀해주신 것 처럼 export default를 바깥에서 하니 제가 봤던 에러가 더이상 발생하지 않네요. 하지만 네.. 왜인지 public variable은 나타나지 않네요. ㅠㅠ 네임스페이스는 사용하지 말아야 하나 봅니다.
답변 정말 감사합니다!

**Reply by mayko1234** [reply]
지금 Puerts코드 보고있는데요, 알려주셔서 넘넘 감사합니다! 
궁금하던 부분들이 있었는데 해결되고 있는 중이예요..ㅎㅎㅎ

**Reply by runasy-koonta** [reply]
대략 추정하는 내용이지만, TS 코드를 작성하면

TS 코드 --> Zepeto Compiler (여기서 TS -> JS로 변환됨) --> puerts로 작동하게 되는 것 같습니다.
코드를 작성한 뒤 프로젝트 창에서 Show Compiled Script 버튼을 누른 후에 표시되는 코드가 Zepeto Compiler에서 컴파일한 코드 같고요,

이 코드는 프로젝트 폴더 내의 Library/PackageCache/zepeto.script@1.1.1/editor/compiler/zepeto-script-compiler.tgz를 압축 해제하여 나온 dist/index.js 파일을 다음과 같이 실행하여 얻을 수 있습니다.
```
node dist/index.js --project [프로젝트 네임] --port [컴파일러 데몬 작동할 port] --file [컴파일할 파일 경로]
```

이렇게 보시면 컴파일된 코드를 확인할 수 있는데, 원본 코드와 비교했을 때 작성한 코드들이 조금씩 바뀌더라고요.
그래서 여기서 수정하면서 public variable들이 숨겨지는건지, 아니면 zepeto script 모듈의 editor dll에서 제대로 불러오지 못 하는 건지... 이 부분은 Zepeto Script compiler의 문제라면 naverz에서 빠르게 해결해줬으면 하는 문제입니다 ㅠㅠ

**Reply by mayko1234** [reply]
우와 넘 흥미롭네요 ㅎㅎ 감사합니다~ 지금 말씀해주신거 해보려고 하는데 [컴파일러 데몬 작동할 port]가 뭔지 모르겠어요. 시간나실때 추가 설명 부탁드려도 될까요?

**Reply by runasy-koonta** [reply]
이 부분은 저도 Zepeto의 Unity module을 디컴파일 해봐야 알 수 있는 부분이라 정확히는 모릅니다 ㅠㅠ

다만 예상하는 것은, Unity에서 Node.js로 만들어진 컴파일러에 어떤 파일을 컴파일할지 데이터를 보내야 할 텐데요,
(TS 파일을 수정한 뒤 Unity Editor에 포커스를 주면 로딩창이 뜨면서 컴파일이 되지요)

이때 HTTP나 TCP 통신을 통해서 컴파일러에 메시지를 주는 게 아닐까 싶습니다.

만약에 Unity Editor에서 해당 컴파일러를 실행시키고, 지정한 포트로 파일명을 HTTP 등으로 전달해주면 해당 파일을 컴파일해주는 것이죠~

그렇기 때문에 테스트 목적으로 해당 스크립트를 실행시킨다면, 단순히 아무 포트나 (물론 시스템 포트로 예약된 1000번 이하의 포트나, 65535를 넘어가는 포트는 안 되겠고, 1234/8888 등의 포트가 적절할 겁니다) 입력해도 문제없이 돌아갑니다.

**TL;DR: 유효한 포트 번호라면 아무 숫자나 입력해도 됩니다. (1234 등)**

**Reply by mayko1234** [reply]
![image](https://user-images.githubusercontent.com/97988822/151538580-50c14999-eb0e-483d-ace8-01684fbe936c.png)
우선 7777넣고 엔터쳤는데, 이 상태에서 멈춰있습니다. ㅠㅠ 계속 기다려야 할까요? 만약에 이 단계가 끝나면 컴파일된 코드는 어디에 저장되어 있을까요?
(답변 넘 감사합니다! )

**Reply by runasy-koonta** [reply]
원래는 정상적이라면 해당 콘솔 창에 표시되는게 정상입니다!
다만 file 부분에는 컴파일할 코드의 위치를 입력해주시면 됩니다. (.ts 인 파일)

**Reply by mayko1234** [reply]
아~! 제가 잘못이해했었네요. 지송합니다. 하핫 말씀해주신 것처럼 수정해서 컴파일된 코드 얻었어요! 언리얼 블루프린트를 코드로 자동 전환 했을때랑 비슷한 느낌의 코드가 나오네요...^^;;;; 넘 재밌어요~! 친절하게 답변해주셔서 다시 한번 감사합니다! 

![image](https://user-images.githubusercontent.com/97988822/151539803-6a83defa-2964-4af2-bab7-e72050780ad4.png)

혹시 제가 위에 잘못 적어놓은 코멘드때문에 헷깔리실 분이 있을까봐 제가 쓴 코맨드도 함께 공유합니다~!

`node C:\Zepeto\NewWorld\node_modules\zepeto-script-compiler\dist\index.js --project NewWorld --port 8888 --file C:\Zepeto\NewWorld\Assets\Scripts\Template\Coroutine.ts`
